### ISSUE TITLE
[distribution] Debian packages and ppa

### ISSUE THREAD

**Issue by vacary** [siconos/siconos#32]
Extension of the work of @xhub for [PPA for siconos i386 build of siconos 3.9~alpha-0~487...](https://launchpad.net/~oli-huber/+archive/ubuntu/siconos/+build/9361605) for Debian.

[https://launchpad.net/debian/+ppas](https://launchpad.net/debian/+ppas) ?

**Comment by xhub** [comment]
It's not what you think, I believe they have just imported all the source packages, but are not building any deb pkg. It may be easier for them to fork things once it is imported on launchpad.

https://build.opensuse.org/ looks promising

**Comment by radarsat1** [comment]
I will high-jack this issue since it is still open.  Recent discussions with OSRF revealed that they would highly prefer integrating Siconos into Gazebo when Siconos is available through standard Debian packages. Suddenly it becomes a priority for me ;)

I suggest making a siconos-debian repository here where we can all contribute to making a solid (set of) Debian package(s) based on the latest release.  (We can do it based on 4.0 but update as soon as 4.1 is released.)  @xhub would you mind uploading your work on the PPA, or do you prefer developing it elsewhere?

The reason I pluralise package(s) above is that I wonder if it would be a good idea to have separate siconos-numerics, siconos-kernel, siconos-mechanics, siconos-mechanics-bullet packages etc.?

This would help handling dependencies.  For example, avoid installing OpenCascade unless the user really wants it.  We would have to agree on ways to configure Siconos options for the packages.. e.g. currently it's not possible to build siconos-mechanics compiled with WITH_BULLET=ON against a siconos-kernel compiled without it.

Ideally the dependencies would be sane, like:

* siconos-numerics requires BLAS, LAPACK etc
* siconos-kernel requires siconos-numerics
* siconos-mechanics requires siconos-kernel
* siconos-mechanics-bullet requires siconos-mechanics and libbullet
* siconos-mechanics-mechanisms requires siconos-mechanics and libOCE

Currently I'm not sure how you could set this up with our current CMake system, since options like WITH_BULLET are global to the whole project.

**Comment by radarsat1** [comment]
I think alternatively if we are not ready to split mechanics like that (understandable, I'm not even sure how to do it..) then we could decide for now that OCC stuff is still in development and we just compile mechanics WITH_BULLET and leave it like that.  The alternative is to make dependencies on bullet and OCC at the same time .. the "just install everything" approach.

**Comment by xhub** [comment]
* Stephen Sinclair <notifications@github.com> [2016-12-19 10:19:45 -0800]:

> I will high-jack this issue since it is still open.  Recent discussions with OSRF revealed that they would highly prefer integrating Siconos into Gazebo when Siconos is available through standard Debian packages. Suddenly it becomes a priority for me ;)

Very good!

> I suggest making a siconos-debian repository here where we can all contribute to making a solid (set of) Debian package(s) based on the latest release.  (We can do it based on 4.0 but update as soon as 4.1 is released.)  @xhub would you mind uploading your work on the PPA, or do you prefer developing it elsewhere?

I created a repository on launchpad because at the time I looked into it, it was not possible to merge a git repo inside another repo. Now it may be possible. I believe that we came to a consensus that we should put a debian directory in the siconos source and ave all the necessary files there.

By luck, I somehow managed to do what I failed a few months back: to have a siconos team project on Launchpad.  I moved the repo to https://code.launchpad.net/~siconos/siconos/siconos-deb I added Maurice and Steve to the team, since I could find their launchpad account. I couldn't find any for neither Vincent nor Franck.

> The reason I pluralise package(s) above is that I wonder if it would be a good idea to have separate siconos-numerics, siconos-kernel, siconos-mechanics, siconos-mechanics-bullet packages etc.?

Yes that was on my todolist. From my limited understanding, it is easy to create siconos-{module} packages, since we just have to select the right files (lib + includes). It may be slightly more challenging for includes, since we don't separate it by module, but a workaround is easy.

It may not be the best example, but you can have a look at the MUMPS debian packaging:  http://sources.debian.net/src/mumps/4.10.0.dfsg-4/debian/

> This would help handling dependencies.  For example, avoid installing OpenCascade unless the user really wants it.  We would have to agree on ways to configure Siconos options for the packages.. e.g. currently it's not possible to build siconos-mechanics compiled with WITH_BULLET=ON against a siconos-kernel compiled without it.

I don't think that's Debian's philosophy. My impression is that they build a fat package, with everything that is ``free'' enabled. If the user is interested in having only minimal dependencies, there are other distro like Gentoo that catered to them.

Splitting the packages would help a bit, if the user only wants numerics or is not interested in the python scripts and therefore does not need vtk ...  I think that right now, most dependencies are reasonable, except for vtk. But the OCE support is not compiled.

Also I don't see your last point: WITH_BULLET does appear in the kernel directory.

> Ideally the dependencies would be sane, like:
> 
> * siconos-numerics requires BLAS, LAPACK etc
> * siconos-kernel requires siconos-numerics
> * siconos-mechanics requires siconos-kernel
> * siconos-mechanics-bullet requires siconos-mechanics and libbullet
> * siconos-mechanics-mechanisms requires siconos-mechanics and libOCE
> 
> Currently I'm not sure how you could set this up with our current CMake system, since options like WITH_BULLET are global to the whole project.

Well, look at the mumps example or another package were they split the installed files (most c/c++ software are packaged that way since they like to have a *-dev pkg)
It's easy to create packages by installed files. If you want to have different builds (Bullet, OCE), that's going to be difficult.

I would enable by default all the library that are not too heavy and available in the standard debian/ubuntu distro. I like very much the idea of splitting the package into siconos-module to lessen the installation burden if the user does not want the full pkg.

Also I think the python files should be separated from the lib and dev packages. We also need to rework the build system to enable multiple python versions (2.7, 3.4, 3.5)  in a single build. But that has a lower priority.

**Comment by xhub** [comment]
> I think alternatively if we are not ready to split mechanics like that (understandable, I'm not even sure how to do it..) then we could decide for now that OCC stuff is still in development and we just compile mechanics WITH_BULLET and leave it like that.  The alternative is to make dependencies on bullet and OCC at the same time .. the "just install everything" approach.

One issue is that OCE is not really available as a debian pkg, isn't it?

I saw that freecad has oce pkgs https://launchpad.net/~freecad-maintainers/+archive/ubuntu/oce-release, but maybe I'm confused.

I could not find how they had this ppa so that the build process can go forward with OCE.

**Comment by radarsat1** [comment]
Ok, I agree on all points I think.  In any case a "fat" package as you mentioned is much easier for us right now, so let's go with that and wait for comments from Debian.  Thanks for adding me to the PPA, I have no idea how to work with launchpad but I will learn.   I thought maybe we could develop the packages in git based on siconos release tarballs, not by embedding the git repository actually.  I was under the impression that Debian packagers often treat a package as a separate project.  But perhaps launchpad is the way to go, I'll try it.  I'll take up the task of proposing it to Debian Science and hopefully getting a DD to take the package, I am fine with it being a long-term effort ;)

As for OCE, I was using the Ubuntu FreeCad-related packages, although I had so much difficulty getting it to work that maybe I am doing something wrong.  I will continue that discussion with Maurice in #84.

**Comment by vacary** [comment]
Really happy that the issue remerges. A good Xmas gift :-)
I agree with all the points. Let us do a fat packages if we are able to get the dependent packages through the debian distro. If not, for one and pythonoce for instance, then we can split the packages or building one with all inside
On ubuntu, all dependencies are provided but the standard distro.  see our rc.ubuntu file. It is not the same for debian ?

**Comment by radarsat1** [comment]
I still have to look into the differences between Ubuntu and Debian, but isn't rc.ubuntu only for the base siconos setup?  (i.e. numerics + kernel)  It doesn't include bullet for example.

**Comment by xhub** [comment]
> Ok, I agree on all points I think.  In any case a "fat" package as you mentioned is much easier for us right now, so let's go with that and wait for comments from Debian.

> Thanks for adding me to the PPA, I have no idea how to work with launchpad but I will learn.   I thought maybe we could develop the packages in git based on siconos release tarballs, not by embedding the git repository actually.  I was under the impression that Debian packagers often treat a package as a separate project.  But perhaps launchpad is the way to go, I'll try it.  I'll take up the task of proposing it to Debian Science and hopefully getting a DD to take the package, I am fine with it being a long-term effort ;)

I had to host the repo siconos-deb on launchpad using bazaar since it would not merge a git repo (the nest-part) in the recipe. Now it appears that it is supported. I'm all for dropping that repo on launchpad and move it to github, but we need to do the import and to modify the recipe.

Launchpad is not the panacea, but it works reasonably well and once the configuration is done, it just works. Plus the ppa is the way to distribute it until it is in Debian. And even then, it is nice to have daily builds for the git/

It's nice that you step up for maintaining it in Debian. I think we still need to create a lot of separate packages (siconos-module, siconos-module-dev, python-siconos-module) before submitting it.  Plus lintian complains a bit and we would have to fix that before submitting it imho.

> As for OCE, I was using the Ubuntu FreeCad-related packages, although I had so much difficulty getting it to work that maybe I am doing something wrong.  I will continue that discussion with Maurice in #84.  

OK. I found how to add PPA dependencies for building, so now I know how to make this work. I'd still think that we should not depend on another ppa for the base pkg (exception made to backports like cmake, cf trusty or precise).

**Comment by xhub** [comment]
* Stephen Sinclair <notifications@github.com> [2016-12-20 08:23:24 -0800]:

> I still have to look into the differences between Ubuntu and Debian, but isn't rc.ubuntu only for the base siconos setup?  (i.e. numerics + kernel)  It doesn't include bullet for example.

I'm not sure rc.ubuntu is really up-to-date. Nothing uses it. A better list of dependencies should be in debian/control.

**Comment by xhub** [comment]
I changed most of the launchpad-related things to be part of the siconos group. The ppa is now at https://code.launchpad.net/~siconos/+archive/ubuntu/siconos-daily it will get slowly populated. I create a ppa for fclib at https://code.launchpad.net/~oli-huber/+recipe/fclib-daily in order to enable fclib in siconos.

**Comment by radarsat1** [comment]
It seems your PPA built successfully, great!  On my end I have been working on a more finely split up set of packages for Debian/Sid.  I have been working in git so far, as it seems the best way to link the package management to the Siconos repository.  Also, I have been working on the principle that these packages will be based on the latest release.  However, I like the idea of the PPA providing a daily build too.  So here's what I suggest:

* siconos-daily remains as it is, always updates to latest git master, it is a totally "fat" package including everything that siconos can build.
* siconos-latest is a separate PPA containing a set of packages splitting up Siconos along module lines that we would propose for inclusion to Debian.  Currently it builds 4.0.0, but can be updated to 4.1.0 on our next release

I think I can push directly to a PPA using "dput" from the git repository.  Actually the git repository is just a clone of siconos with an added "debian" folder.  Theoretically we could do this just in a branch of the main siconos repository, or I could just keep it separate.  (We could create github/siconos/siconos-deb)

I have taken extensive notes during the development so I'll add those to the repository.

One thing I don't understand is where in your PPA it says that it should pull in the sources?  I see the Vcs-Git line, but nothing telling it what version or branch to use..
I guess I was surprised not to see the sources in the bzr repository, so I'm not sure where the instructions to grab the sources from github are.

Oh, as for fclib, I am not sure whether to include it or not.  Ideally components of siconos-numerics refering to fclib could be separated into siconos-fclib, but I'm not sure if that's possible.

**Comment by radarsat1** [comment]
I should mention that I tried to make a log-in on Debian Science [alioth](https://alioth.debian.org/projects/debian-science/), because I am under the impression that maybe we can put a git repository for the package there and it might help when asking for sponsorship.. However, I was unable to make the account.  I got the verification email and when I clicked on the link it said my user name does not exist.  Perhaps I'll have to ask the Debian Science team for the steps to take once we are happy with the package.

**Comment by xhub** [comment]
> * siconos-daily remains as it is, always updates to latest git master, it is a totally "fat" package including everything that siconos can build.
> * siconos-latest is a separate PPA containing a set of packages splitting up Siconos along module lines that we would propose for inclusion to Debian.  Currently it builds 4.0.0, but can be updated to 4.1.0 on our next release

I'm not sure I'd separate the two. Ideally, what goes into siconos-latest is a snapshot of the debian directory used for siconos-daily and updated so that the package name is correct.
Ideally the user can switch between the two repository easily. For instance if we fix a bug in the master version or implement a new feature, he should be able to switch to the siconos-daily packages.
We are not making releases frequently so it should be easy to try the git version.

I'm also not sure we have the same definition of "fat" packages.

> I think I can push directly to a PPA using "dput" from the git repository.  Actually the git repository is just a clone of siconos with an added "debian" folder.  Theoretically we could do this just in a branch of the main siconos repository, or I could just keep it separate.  (We could create github/siconos/siconos-deb)

I'm all for creating github.com/siconos/siconos-deb

I still think that it may be a good thing to keep building the pkgs on launchpad. It's quite hassle free and we can support many distro very easily.

> One thing I don't understand is where in your PPA it says that it should pull in the sources?  I see the Vcs-Git line, but nothing telling it what version or branch to use..
> I guess I was surprised not to see the sources in the bzr repository, so I'm not sure where the instructions to grab the sources from github are.

It uses an import from the master branch done by the people in hrp: https://code.launchpad.net/~hrg/hrg-packaging/siconos The recipe siconos-daily (https://code.launchpad.net/~siconos/+recipe/siconos-daily) is used to create the source deb. Then it gets compiled. It would still be good to get the handle on that import: Launchpad does not allow two import of the same git repo.

> Oh, as for fclib, I am not sure whether to include it or not.  Ideally components of siconos-numerics refering to fclib could be separated into siconos-fclib, but I'm not sure if that's possible.

Why not? I created another ppa for it and the packaging is rather straightforward. Actually I'm working on the packaging again since a grad student here wants to run the fc3d solver in siconos that uses his solver. We use fclib to have interesting problems to solve.

**Comment by radarsat1** [comment]
I realize maybe you think that I'm over-complicating this, so I should
clarify my reasons for wanting to split the package up.  As you know, a lot
of packages in Debian are split into the library binaries and a -dev
package containing the headers, and sometimes a -tools package containing
executables for /usr/bin.  The reason for this is quite clear -- installing
programs that just use the library should not require installing all the
headers and everything.

In the case of Siconos, historically we use it as a whole simulation
framework, using the "siconos" executable to compile and run simulations
defined in C++ or Python, which requires the whole package, including
headers, to be available.  However, in the context of my project of using
Siconos as a physics engine, I'll really be using Siconos as a C++ library
only.  In that sense it is reasonable to have a "libsiconos" package that
only installs the .so files that dependent projects need to run.  This will
become more important if Siconos picks up usage as a rigid body engine.

I also think Debian will be more likely to pick up the package proposal if
we split it up along traditional lib/dev lines.  In general the package
should be based on release versions only, so that projects using it can
rely on the API from a specific version.  However, of course as you say it
should be easy to switch to the daily build for work that needs the latest
stuff.  I believe this can be done easily by setting up the daily PPA to
declare conflicts with the "latest release" PPA.

I'm not sure I'd separate the two. Ideally, what goes into siconos-latest
> is a snapshot of the debian directory used for siconos-daily and updated so
> that the package name is correct.
> Ideally the user can switch between the two repository easily. For
> instance if we fix a bug in the master version or implement a new feature,
> he should be able to switch to the siconos-daily packages.
> We are not making releases frequently so it should be easy to try the git
> version.
>

In Debian time actually a release every 6 months or so is pretty good.  But
in any case, if Conflicts / Replaces are done properly, I don't think
switching to siconos-git would ever be more difficult than just adding the
siconos-daily ppa and issuing "apt-get install siconos-git".  This would
automatically replace all siconos-latest packages with the "fat" git master
version.  So I don't see a problem with organizing the two sets of packages
differently, it's just a matter of setting these fields correctly.


>
> I'm also not sure we have the same definition of "fat" packages.
>

We should clarify then ;)

For me, I assumed "fat" package meant a single package that installs
everything you get from "make install" with as many configuration options
enabled as possible.

As opposed to having slim, targeted packages like siconos-numerics, that
contains only /usr/lib/*/libsiconos_numerics.so.


>
> > I think I can push directly to a PPA using "dput" from the git
> repository. Actually the git repository is just a clone of siconos with an
> added "debian" folder. Theoretically we could do this just in a branch of
> the main siconos repository, or I could just keep it separate. (We could
> create github/siconos/siconos-deb)
>
> I'm all for creating github.com/siconos/siconos-deb
>
> I still think that it may be a good thing to keep building the pkgs on
> launchpad. It's quite hassle free and we can support many distro very
> easily.
>

Yes, I need to continue to familiarize myself with launchpad.  Maybe it's
even possible to have it automatically mirror siconos-deb?  I'll check into
it.


> It uses an import from the master branch done by the people in hrp:
> https://code.launchpad.net/~hrg/hrg-packaging/siconos The recipe
> siconos-daily (https://code.launchpad.net/~siconos/+recipe/siconos-daily)
> is used to create the source deb. Then it gets compiled. It would still be
> good to get the handle on that import: Launchpad does not allow two import
> of the same git repo.
>

Ah thank you, that sort of clarifies things.  I'll read up on "recipes" !

Why not? I created another ppa for it and the packaging is rather
> straightforward. Actually I'm working on the packaging again since a grad
> student here wants to run the fc3d solver in siconos that uses his solver.
> We use fclib to have interesting problems to solve.


It's a good idea, we can propose fclib for Debian too maybe ;)

I just meant that I wasn't sure if it's possible to generate a
libsiconos_numerics.so that refers to fclib but also works without it..
I mean, if the user wants to install libsiconos_numerics and doesn't care
about fclib, he shouldn't be forced to install it.
I'll keep experimenting.

I don't know much about fclib, so to be honest I don't understand why it's
a thing that siconos depends on, instead of the other way around.  If it's
a collection of friction contact problems, shouldn't it be calling the
siconos solvers?  Instead it seems that siconos calls fclib.

Ok, looking at the source, it seems that fclib is called for two things:
(1) writing the problems out in a compatible file format, and (2) for
reading them to use as Siconos test cases.

(1) dependency could be broken by just including the writing code into
Siconos.
(2) is only needed during testing.

So I would propose that a dependency on fclib is useful for the git daily
build, but not so useful for Debian packages.
Perhaps it could be useful for a siconos-examples package.

**Comment by xhub** [comment]
Disclaimer: I'm quite jetlagged, so some things may not make sense.

I think we have the same goal: to have a packages that conform to the
debian standards (and also rpm-based distro have a similar philosophy),
and minimize, if possible, the external dependencies that the user have
to install even tho he does not use it (I'm looking at you, python-vtk!)
My colleague just uses siconos-numerics and has to use 1Go of HDD
because we have a full pkg only.

I would say that my feeling is that we should enable all the
dependencies that are in the std repo, but try to split the siconos pkg
as much as possible so as to minimize what the user has to install for
his need.

This is not ``over-complicated'', it's just the right way imho. And the
effort is not only for debian-based distro, but also for the ones based
on rpm.

I want the same splitting for the git based builds. I did not do it since it
was already time consuming to make siconos compile as a debian package
and do some preliminary QA.

Now what is left to be done is to create multiple .deb from the build.
The major blocker appears to be the header files, since they are all
installed in /usr/include/siconos. It is then difficult to select the
ones that should be installed in siconos-control-dev, for instance.
An easy fix is to install the headers in /usr/include/siconos/$module,
but that's not a decision I want to make by myself. But I think that's a
good idea in any case.

> I realize maybe you think that I'm over-complicating this, so I should
> clarify my reasons for wanting to split the package up.  As you know, a lot
> of packages in Debian are split into the library binaries and a -dev
> package containing the headers, and sometimes a -tools package containing
> executables for /usr/bin.  The reason for this is quite clear -- installing
> programs that just use the library should not require installing all the
> headers and everything.

Yeah, we may want to do something about /usr/bin/siconos*. In that case,
we could make it depend on gcc instead of any siconos-kernel.

> In the case of Siconos, historically we use it as a whole simulation
> framework, using the "siconos" executable to compile and run simulations
> defined in C++ or Python, which requires the whole package, including
> headers, to be available.  However, in the context of my project of using
> Siconos as a physics engine, I'll really be using Siconos as a C++ library
> only.  In that sense it is reasonable to have a "libsiconos" package that
> only installs the .so files that dependent projects need to run.  This will
> become more important if Siconos picks up usage as a rigid body engine.

Good point. 

> I also think Debian will be more likely to pick up the package proposal if
> we split it up along traditional lib/dev lines.  In general the package
> should be based on release versions only, so that projects using it can
> rely on the API from a specific version.  However, of course as you say it
> should be easy to switch to the daily build for work that needs the latest
> stuff.  I believe this can be done easily by setting up the daily PPA to
> declare conflicts with the "latest release" PPA.

I still think that the best solution is to have the same format. I don't want the user of siconos-numerics to have to install python-vtk just because he want to try the latest version of the code.

> Yes, I need to continue to familiarize myself with launchpad.  Maybe it's
> even possible to have it automatically mirror siconos-deb?  I'll check into
> it.

Yes we should host the debian related files in github and do an automatic import on launchpad.

> It's a good idea, we can propose fclib for Debian too maybe ;)

That would be good.

> I just meant that I wasn't sure if it's possible to generate a
> libsiconos_numerics.so that refers to fclib but also works without it..
> I mean, if the user wants to install libsiconos_numerics and doesn't care
> about fclib, he shouldn't be forced to install it.

While I agree a lot with this principle, this is not how debian and rpm based distro work. They usually enable everything and make the user install package that he will never use. That's how you get an X install on an headless machine. Other distribution, like Gentoo, offer the flexible to choose what gets built and therefore, give the user the choice to enable/disable some dependencies.

> I don't know much about fclib, so to be honest I don't understand why it's
> a thing that siconos depends on, instead of the other way around.  If it's
> a collection of friction contact problems, shouldn't it be calling the
> siconos solvers?  Instead it seems that siconos calls fclib.

It is a library to read/write FC problems. It is solver-independent. It was created to promote data exchange.

> Ok, looking at the source, it seems that fclib is called for two things:
> (1) writing the problems out in a compatible file format, and (2) for
> reading them to use as Siconos test cases.
> 
> (1) dependency could be broken by just including the writing code into
> Siconos.
> (2) is only needed during testing.

No, there exists a collection of problems (fclib-library: https://gforge.inria.fr/frs/?group_id=2824&release_id=8505#fclib-library-of-problems-v0.2-title-content) that are used for doing comparisons. So having fclib in debian and linking to is a good thing I believe. Plus the .dev is 35 kb

Also the purpose of that library is to have consumer and not be embedded in the code.

**Comment by radarsat1** [comment]
Okay, it seems we are agreeing well, perhaps I had misunderstood your
previous comments about the fat package and your intentions for the daily.

And thanks for the clarification on fclib.  I agree, may as well make it a
dependency then, it's tiny anyways.

Actually I've got a test package almost ready with many things split off
from the installed files.  You're right that the headers are a sticking
point, so currently working on a solution for that.  Currently the way I
have it though, it's already possible to install libsiconos and
libsiconos-dev and get only the *.so and header files, which totals only a
couple of megabytes.

The siconos-examples is actually the big one (~26 MB compressed), because
of all the ref files.  Perhaps we can come up with a strategy for that,
like siconos-basic-examples and siconos-full-examples or something..

Excluding examples and debug symbols though, it all adds up to under 3 MB.
I also have a "siconos" package that depends on all the subpackages, so the
whole thing can be installed in principle by "apt-get install siconos".

Can we make a fork of "siconos.git" called "siconos-deb.git" then, in
github?  (I think I still don't have permission to do so...)  Then I can
upload it there.  Alternatively I just put it on the main repo under a
branch called "debian/sid".


I still think that the best solution is to have the same format. I don't
> want the user of siconos-numerics to have to install python-vtk just
> because he want to try the latest version of the code.


Fair enough.  Perhaps once I upload my "control" file we can change
siconos-daily to copy it, and I'm sure you'll have changes or suggestions
at that point.

Yes we should host the debian related files in github and do an automatic
> import on launchpad.


Perfect.  Plus any changes needed for Ubuntu I guess, if any.  I guess if
it's in a new git repository (a fork of siconos.git) then launchpad will
let us do a new import.

**Comment by vacary** [comment]
Cool. It seems that you are converging. I agree with all the points. it is also a good idea to split the include dir, but if you can avoid this nasty work, I would understand.

Normally, @radarsat1 can create as many forks as you want since you are  in the core group (owner and admin). Nevertheless, I would prefer the idea of a branch

**Comment by radarsat1** [comment]
Update:

I added siconos-deb repository.. While working on this I found a few good reasons to keep a separate repository, one is that I needed to add a branch for each supported distro, which could quickly get complicated as there is a new Ubuntu release every 6 months.  The other is that I use a different email address for Launchpad, so this just makes it a lot easier.  (The email address is GPG signed so it's a necessary burden.)

Anyways, siconos-deb contains branches "debian/sid" and "ubuntu/xenial".  I got debian/sid building properly in a Docker container, but am still getting a few build errors with Xenial.  (Mostly related to sphinx for the docs.)  I've been working with a [test PPA on Launchpad](https://launchpad.net/~radarsat1/+archive/ubuntu/siconos-test-ppa) but it still needs work.  When I'm satisfied with it I'll move it to the ~siconos team.

I split it up into quite a few packages, probably the dependencies order is not quite right. I'll post something next week explaining the logic of how I decided to split things up, and I'll certainly be interested in any comments on that.

**Comment by radarsat1** [comment]
Currently waiting for PPA results, I think I've got it now.  The docs proved more difficult than expected to build due to some missing Python packages.  I'll probably leave them out of the Debian build, but I got it working for the Xenial PPA.  Here is the dependency structure I'm proposing, not sure if it's 100% correct, any comments are appreciated:

![image](https://cloud.githubusercontent.com/assets/66805/21486472/16822644-cb94-11e6-9f74-31b1bb89b82e.png)

Packages are:

    siconos: this installs everything needed to run the "siconos" executable, i.e. all -dev and Python packages.
    libsiconos-*..: each Siconos component gets its own library, therefore its own deb
    libsiconos-*-dev: headers for that library (determined based on which directories the headers are in)
    python-siconos: just whats needed for all Siconos Python modules.. I did not split them up by component.
    siconos-mechanics: the "vview" and "vexport" etc tools, and their dependencies (i.e. python-vtk6)
    siconos-doc: just docs, does not depend on anything else, but to solve some lintian problems I removed some internal js and replaced it with Ubuntu package dependencies
    siconos-examples: just the examples, but it depends on python-siconos lots of things get installed, plus matplotlib. I should probably add a gnuplot dependency too.

I think I am missing possibly many run-time dependencies of the libraries, need to do more testing.
Lots more lintian issues to resolve.

**Comment by radarsat1** [comment]
Ok as of last night I now have a [PPA containing debs with no lintian errors or warnings](https://launchpad.net/~radarsat1/+archive/ubuntu/siconos-test-ppa)!

So apart from figuring out what to do with siconos-externals (#95), tasks I have left are the following:

* import the git repository into bazaar in an intelligent way; learn about launchpad "recipes"
* move the ppa over to ~siconos
* adjust the "daily" ppa to use the same package structure; I see four possibilities here,
  i. There are two PPAs, user switches PPA and automatically gets packages with higher version numbers that conflict with 'latest release' packages.
  ii. The 'daily' and 'release' packages are hosted in the same PPA and user chooses by installing a different package name; both 'latest' and 'daily' use the same "Provides" and "Conflicts" packages so that they cannot be installed at the same time.
  iii. The are in the same PPA and use different package names but simply "Conflict" with each other and "Provide" the same package -- `apt-get install libsiconos-mechanics-daily` would automatically uninstall `libsiconos-mechanics0`, something like that?
  iii. They are in the same PPA and use different names and can be installed simultaneously; user chooses at build time which one he wants to compile against.  This would be difficult when considering the executables in /usr/bin/ (siconos, vview, etc.) and could in general be confusing, so I prefer avoid this choice.
* port the Xenial package to Debian/Sid and propose it to Debian.  (need to figure out #95 first..)
* provide in the PPA also packages for other versions of Ubuntu

For options (i) and (ii) above, we need a way to make sure the "daily" package always has a higher version number.  I'll look into how other projects do this.

## On the soname

I had to set the "soname=0" a.k.a the "so-version" or API version of Siconos.  This number gets appended to the package name, i.e. `libsiconos-mechanics0`.  Actually I'm a bit confused since usually this is a 3-part version number but then the name would have be to something like `libsiconos-mechanics0.0.0` which seems rather redundant.  Since we are pretty much always breaking the API (for now) I propose just setting it to 0, and incrementing to 1 when 4.1.0 is released.  I set it in `debian/rules` for now, modifying `cmake/LibraryProjectSetup.cmake` but in fact we should do this in the project code eventually, and each component could have its own so-version which only needs to increase if that component has been touched.  e.g., if libsiconos-control has not changed from 4.0.0 to 4.1.0 then there is no reason to change its so-version.  However, we must remember to increment this on each release when appropriate!

**Comment by radarsat1** [comment]
Forgot, another to-do item is to look into adding a dependency on the fclib package and look into support OpenCascade or not.  These are not so different from figure out #95, in case we end up having to propose several separate packages to Debian.

**Comment by xhub** [comment]
Any updates?

**Comment by radarsat1** [comment]
Not at the moment, been working on mechanics issues for now.

**Comment by xhub** [comment]
Ping again :)

**Comment by radarsat1** [comment]
It is essentially ready to be proposed to debian science after the next release. I did not yet make an fclib package but I am looking into it.

**Comment by xhub** [comment]
Cool! I'd like to continue building a daily snapshot of the git master
on launchpad, I have to look into your work.

-- 
Olivier

* Stephen Sinclair <notifications@github.com> [2018-03-14 20:51:59 +0000]:

**Comment by radarsat1** [comment]
Status update, it seems that the current round of API cleanup has wound down which is great.  The package builds fine and all tests are passing during the build.  I need to fix some things in the copyrights file.  Other than that, I was sort of blocked and deliberating on three issues.

1. Debian still sadly does not have a python3-compatible VTK package.  They are working on it, I've seen the messages in the mailing list, but it doesn't look "soon".  I spent some time trying to see if I could compile the code for python2 and python3 in separate builds and combine them, but this doesn't seem feasible.  So I suggest (lamentably) removing vview/vexport from the package until python3-vtk is ready.  The only other possibility is to build the package for Python 2 only, but for proposing a new package this seems counterproductive.  A more important goal is to get the libraries in there, python tools notwithstanding.

2. There was a missing package necessary for compiling the documentation, but it appears to be available now so I am working on siconos-doc now.

3. Examples, I don't know whether to package all the CAD files, etc.  I could potentially just drop the examples package.  I'll ask on the mailing list what the preferred thing is to do.  I certainly want to avoid including non-working examples because I removed some files they use.

Also, just to mention, I have all options turned on (serialization, etc) with the exception of `WITH_OCC` and `WITH_MECHANISMS`.  This is because as far as I know the OCC stuff cannot currently be made to run with the current Debian packages.  Please correct me if that's not true.  Like VTK, I see on the mailing list that there is some activity regarding OpenCascade, so hopefully this will be resolved.

I don't see much that can necessary block a Siconos release, but give me a bit of time to ensure the packaging is in 100% good shape so that we can synchronize the release with the package proposal.

And @xhub, I will re-init the git repository at that time and push it to salsa so that you can take a look.  I think it might even be possible to integrate it into the Siconos CI at Inria.  I have a Docker file that can be used to clone the package repository inside a "Debian unstable" container and build the debs.  (Unfortunately I haven't got autopkgtest working under Docker.)  If you want to use it for the daily snapshot builds it might be possible to adapt it but you'll need to fork the debian/changelog with git-versioned updates, so I don't know how complicated that could become, but it might be feasible.  Otherwise I see no problem just having a one-package-for-everything version for the daily build as you have now, just make it "conflict" with all the Debian packages.

A final issue is that I still need to add some "as-packaged" testing (called autopkgtest) to ensure that various aspects of Siconos work after the package is installed.  I am also still waiting for fclib to be accepted, so far after a couple of days no sponsors have shown interest.

For information the particulars of the package build: no RPATH, tests must run prior to install (hence my recent fixes with regards to LD_LIBRARY_PATH), Python 3-only, serialization enabled, system suitesparse.

The only patch I apply, which I have not pushed here, is something that breaks up the installed script `siconosTargets.cmake` into components.  That is, currently every component that Siconos was compiled with must be present or cmake will fail, but in the packaging scenario it's possible to compile kernel but only install numerics, for example.  I don't know if that's interesting to contribute in the project, or if it's too specific to packaging, but it's a fairly minimal change.

**Comment by bremond** [comment]
>So I suggest (lamentably) removing vview/vexport from the package until
>python3-vtk is ready.
Anyway, the main goal is to have a package for Gazebo integration and vview,
vexport are not needed for this.

for OCC it is the same it is not needed. oce (OpenCascade community
edition) is already packaged but pythonocc is not and we rely on it.

On my ubuntu 17.10:
>aptitude search liboce
p   liboce-foundation-dev                                          - OpenCASCADE Community Edition CAE platform library development files
[...]

**Comment by xhub** [comment]
Thanks for the update. I'll try to have a look at the packaging.

**Comment by radarsat1** [comment]
Right, I forgot about that.  I wonder if pythonocc would be hard to package.  I guess I wouldn't try without consulting with the OCE packager, but it could be possible.  It seems Python packages are not too difficult to build.  No promises yet but I can look into it.  Perhaps I'll ask on the mailing list if there is any interest.

**Comment by radarsat1** [comment]
You can [see the package here](https://salsa.debian.org/science-team/siconos), I have set it up for the Salsa Gitlab CI but I am unable to use it to compile the package with serialization and documentation turned on.  (Too long build, too long log, OOM problems. It is not an industrial strength CI system.)

In any case it [passes build and autopkgtest stages with these disabled.](https://salsa.debian.org/science-team/siconos/-/jobs/65462)

I need to create the symbols files and file the Intent to Package, then I will ask for feedback to the community.  I am thinking about whether to make a doc package or not, but I might skip it for now.  I still need to build with docs however in order to have Python docstrings.

**Comment by xhub** [comment]
Great news. Do you need the full doc or just the doxy2swig?

**Comment by radarsat1** [comment]
If I don't build a doc package then I just need doxy2swig, but that implies USE_DOCUMENTATION=ON does it not?  I haven't had time to get back into it for a few days, I'll dive in again soon and do the symbols stuff.  I can't believe how badly USE_SERIALIZATION=ON destroys the CI system (and my laptop), it seems to be quite a bit better with clang.  Well, all the more reason to provide a precompiled package.

**Comment by radarsat1** [comment]
Good news, the siconos package is [now in the NEW queue](https://ftp-master.debian.org/new.html)

**Comment by vacary** [comment]
Now it is in the unstable distribution of debian. Thank you very much @radarsat1
